nvflare/app_common/ccwf/server_ctl.py

```diff
@@ -0,0 +1,459 @@
+        for c in self.participating_clients:
+            reply = resp.get(c)
+            if not reply:
+                self.log_error(fl_ctx, f"not reply from client {c} for ending workflow {self.workflow_id}")
+                num_errors += 1
+                continue
+
+            assert isinstance(reply, Shareable)
+            rc = reply.get_return_code(ReturnCode.OK)
+            if rc != ReturnCode.OK:
+                self.log_error(fl_ctx, f"client {c} failed to end workflow {self.workflow_id}: {rc}")
+                num_errors += 1
+
+        if num_errors > 0:
+            self.system_panic(f"failed to end workflow {self.workflow_id} on all clients", fl_ctx)
+
+        self.log_info(fl_ctx, f"Workflow {self.workflow_id} done!")
+
+    def process_task_request(self, client: Client, fl_ctx: FLContext):
+        self._update_client_status(fl_ctx)
+        return super().process_task_request(client, fl_ctx)
+
+    def process_config_reply(self, client_name: str, reply: Shareable, fl_ctx: FLContext) -> bool:
+        return True
+
+    def _process_configure_reply(self, client_task: ClientTask, fl_ctx: FLContext):
+        result = client_task.result
+        client_name = client_task.client.name
+
+        rc = result.get_return_code()
+        if rc == ReturnCode.OK:
+            self.log_info(fl_ctx, f"successfully configured client {client_name}")
+
+            try:
+                ok = self.process_config_reply(client_name, result, fl_ctx)
+                if not ok:
+                    return
+            except:
+                self.log_error(
+                    fl_ctx, f"exception processing config reply from client {client_name}: {secure_format_traceback()}"
+                )
+                return
+            cs = self.client_statuses.get(client_name)
+            if cs:
+                assert isinstance(cs, ClientStatus)
+                cs.ready_time = time.time()
+        else:
+            error = result.get(Constant.ERROR, "?")
+            self.log_error(fl_ctx, f"client {client_task.client.name} failed to configure: {rc}: {error}")
+
+    def client_started(self, client_task: ClientTask, fl_ctx: FLContext):
+        return True
+
+    def _process_start_reply(self, client_task: ClientTask, fl_ctx: FLContext):
+        result = client_task.result
+        client_name = client_task.client.name
+
+        rc = result.get_return_code()
+        if rc == ReturnCode.OK:
+            try:
+                ok = self.client_started(client_task, fl_ctx)
+                if not ok:
+                    return
+            except:
+                self.log_info(fl_ctx, f"exception in client_started: {secure_format_traceback()}")
+                return
+
+            self.cw_started = True
+        else:
+            error = result.get(Constant.ERROR, "?")
+            self.log_error(
+                fl_ctx, f"client {client_task.client.name} couldn't start workflow {self.workflow_id}: {rc}: {error}"
+            )
```

Comment:
this method similar to `_process_configure_reply` maybe we can consolidate them

Reply:
Let's not do that too soon since things are still early - they can change to become different.